### PR TITLE
Re-core some String APIs on swift-foundation

### PR DESCRIPTION
**Pull request by jmschonfeld**
This removes some String APIs in SCL-F in favor of those in swift-foundation and removes some `!DEPLOYMENT_RUNTIME_SWIFT` paths that are no longer relevant in this project. This depends on https://github.com/apple/swift-foundation/pull/739, so I'll update the commit hash in the package manifest and cmake files once that is merged before merging this

### CODE REVIEW

File: Tests/Foundation/TestDataURLProtocol.swift [hunk 1]
```diff
@@ -76,9 +76,9 @@ class TestDataURLProtocol: XCTestCase {
             ("data:;charset=utf-16;base64,2D3caCAN2D3caCAN2D3cZyAN2D3cZw==", "👨‍👨‍👧‍👧", (expectedContentLength: 22, mimeType: "text/plain", textEncodingName: "utf-16")),
             ("data:;charset=utf-16le;base64,Pdho3A0gPdho3A0gPdhn3A0gPdhn3A==", "👨‍👨‍👧‍👧", (expectedContentLength: 22, mimeType: "text/plain", textEncodingName: "utf-16le")),
             ("data:;charset=utf-16be;base64,2D3caCAN2D3caCAN2D3cZyAN2D3cZw==", "👨‍👨‍👧‍👧", (expectedContentLength: 22, mimeType: "text/plain", textEncodingName: "utf-16be")),
-            ("data:application/json;charset=iso-8859-1;key=value,,123", ",123", (expectedContentLength: 4, mimeType: "application/json", textEncodingName: "iso-8859-1")),
+//            ("data:application/json;charset=iso-8859-1;key=value,,123", ",123", (expectedContentLength: 4, mimeType: "application/json", textEncodingName: "iso-8859-1")),
```
Comment (jmschonfeld) [inline Tests/Foundation/TestDataURLProtocol.swift:79]: These are temporarily disabled until we can enable the `isoLatin1` encoding in swift-foundation or add an up-call to SCL-F to handle the extra encodings beyond the standard UTF encodings

File: Sources/Foundation/NSStringAPI.swift [hunk 2]
```diff
@@ -297,6 +212,7 @@ extension String {
     /// frees the buffer.
     ///
     /// - Warning: This initializer is not memory-safe!
+    @available(swift, deprecated: 6.0, message: "String does not support no-copy initialization")
```
Comment (jmschonfeld) [inline Sources/Foundation/NSStringAPI.swift:215]: These (and the other deprecated functions) are now deprecated on Darwin, so this updates these to match Darwin

File: Sources/Foundation/NSStringAPI.swift [hunk 3]
```diff
@@ -14,58 +14,12 @@
 //
 //===----------------------------------------------------------------------===//
 
-// Important Note
-// ==============
-//
-// This file is shared between two projects:
-//
-// 1. https://github.com/apple/swift/tree/master/stdlib/public/Darwin/Foundation
```
Comment (jmschonfeld) [inline Sources/Foundation/NSStringAPI.swift:22]: This file no longer exists / this is no longer the case, so we can remove this comment altogether

**Comment by jmschonfeld** [comment]
@swift-ci please test Linux platform

**Comment by jmschonfeld** [comment]
@swift-ci please test Windows platform

**Comment by jmschonfeld** [comment]
@swift-ci please test Linux platform

**Comment by jmschonfeld** [comment]
@swift-ci please test Linux platform

**Review by parkera — APPROVED** [review]

**Pull request merged**